explainers/subresource-loading.md

```diff
@@ -0,0 +1,101 @@
+
+- Caching stories with JS bundlers (e.g. webpack) is tough. They are pretty good
+  tools but configuring them to work in an optimal way is tough, and sometimes
+  also conflicts with new requirements like dynamic bundling (e.g. small edit
+  with tree shaking could invalidate everything).
+
+- With JS bundlers, execution needs to wait for the full bytes to come. Ideally
+  loading multiple subresources should be able to utilize full streaming and
+  parallelization, while it's not possible if all resources are bundled as one
+  javascript. (For JS modules execution still needs to be waited for the entire
+  tree due to the current
+  [deterministic execution model](https://docs.google.com/document/d/1MJK0zigKbH4WFCKcHsWwFAzpU_DZppEAOpYJlIW7M7E/edit#heading=h.5652gd5ks5id))
+
+- Related issues: [#411](https://github.com/WICG/webpackage/issues/411),
+  [#526](https://github.com/WICG/webpackage/issues/526)
+
+## Proposal
+
+We propose a new `link rel=”webbundle”` notion [**1**] that lets the browser
+fetch bundled resources and attach them to the page. Such as:
+
+```html
+<link
+  rel="webbundle"
+  href="https://example.com/dir/subresources.wbn"
+  scope="https://example.com/scope/"
+/>
+```
+
+If the main document has such a `link` element, a browser must do:
+
+- A browser must fetch the specified Web Bundle, `subresources.wbn`.
+
+- All the subresources whose URLs are under the scope
+  `https://example.com/scope/` must [**2**] be loaded from the specified Web
+  Bundle.
+
+[**1**] The syntax is tentative. The API surface is not so important for us yet.
+Any alternative ideas are welcome. This is basically about mapping a set of
+resources to a particular bundled resource. Something like
+[Fetch maps proposal](https://discourse.wicg.io/t/proposal-fetch-maps/4259)
+might be able to extend to this case.
+
+[**2**] The exact behavior is TBD. There are several possible behaviors when the
+bundle does not contain the subresource; e.g. just fail, fallback to the
+network, or allow users to specify its behavior with a `link` element's
+attribute.
+
+## Example
+
+### The bundle
+
+Suppose that the bundle, `subresources.wbn`, includes the following resources:
+
+```
+- https://example.com/scope/a.js (which depends on ./b.js)
+- https://example.com/scope/b.js
+- https://example.com/scope/c.png
+- … (omitted)
+```
+
+### The main document
+
+```html
+<link rel="webbundle"
+    href="https://example.com/dir/subresources.wbn"
+    scope="https://example.com/scope/"
+/>
+
+<script type=”module” src=”https://example.com/scope/a.js”></script>
+<img src=https://example.com/scope/c.png>
+```
+
+Then, a browser must fetch the bundle, `subresources.wbn`, and load
+subresources, `a.js`, `b.js`, and `c.png`, from the bundle.
+
+## Subsequent loading and Caching
+
+On the subsequent navigations for the same page/origin, it’s desirable that the
+resources that are already fetched in the bundle shouldn't be refetched. We're
+thinking about two possibilities here:
```

Comment:
Both of those option sound like exclusion-based fetching. What we've observed while working on a similar system is that there are scale boundaries where it stops working. When the number of individual items is high, it breaks the length limits for URLs (1) and a digest becomes increasingly unreliable (2).

The third solution is to have an official way to batch resources by inclusion instead. It requires an "inclusion digest" or "inclusion list" instead of an "cached/exclusion digest" though.

Reply:
@jkrems
Thanks! I agree that especially for dynamic bundling it seems to need some "inclusion signal" in requesting phase. I suspect it still could be benefited if it's combined with cache/exclusion approach for better cache efficiency?  A straw-idea could be:

1. The bundle request should express some hash or list that indicates the set of resources it wants.  Possibly the server and client can share a logic to compute the hash / variations for the session to determine what resources are needed.
2. Platform also attaches some digest that indicates exclusive-fetching information for caching.

This explainer didn't tell anything about 1. but probably it'd be better to mention it too. Does it make better sense, or I might be still missing something?

Reply:
Makes sense, I think you covered what I had in mind! And this definitely doesn't have to block this initial PR, more a comment towards future iteration.